### DISCUSSION TITLE
Horizontally scalable Dragonfly Cluster

### DISCUSSION THREAD

**Discussion by fdsmax** [dragonflydb/dragonfly#1498] [Q&A]
Does Dragonfly support configuring multi-node horizontally scalable cluster (similar to Redis Cluster) for data intensive in-memory cache? Any references on how to configure it with any external tools?

**Comment by chakaz** [comment]
@thulyatech I'm sorry to say that, at the moment, we do not directly support multi-node clusters. If you'd like to set one up, you'll have to program your clients to know which node to connect to for which keys.
That is of course not to say that we do not plan to support clusters in the future, but we're just not there yet.

**Comment by tittuvarghese** [comment]
How Dragonfly is taking care of availability ?

**Reply by chakaz** [reply]
@tittuvarghese Dragonfly supports replication, that is having a master node that streams all of its data (continuously) to another replica node. Upon failures, the replica can be promoted to a master (which means it will stop receiving data from the previous-master node, and will allow writes).
This is the same as Redis, and in fact we also support Dragonfly as a replica of a Redis master.
Does this answer your question?

**Comment by ashotland** [comment]
Hi, so a couple of notes regarding high availability: 
- Dragonfly supports Redis Sentinel - so you can get HA with a master + replica + sentinel deployment 
- We also released a k8s operator https://github.com/dragonflydb/dragonfly-operator so you can get HA on k8s
- I'll let @chakaz chime in on the timeline for cluster, however is important to note that it won't provide HA out of the box. We hope the community will contribute a control plane component that manages a Dragonfly cluster.

Hope that helps :)

**Reply by brandonzylstra** [reply]
@alexted I think the key point there was that HA was not provided _**out of the box**_, and not that it wouldn't support high availability at all.

**Comment by githubzhangfei** [comment]
Can Redis Sentinel be used to manage dragonflydb? That is to say, he supports master-slave replication, but does not support fail over and requires the use of proxies to ensure high availability. Is my understanding correct?

**Reply by ashotland** [reply]
https://github.com/dragonflydb/dragonfly/discussions/1498#discussioncomment-6370531

**Comment by cyppe** [comment]
On this topic - is it supported and/or recommended when running dragonfly operator with for example "replicas: 3", to use the passive "slaves" as read only replicas?

I am not sure if it makes sense, maybe the master will never become a bottleneck in read performance. But just curious if it's recommended or not and if so if there is any service/hostname exposed by the operator that prioritizes connection to slaves in similar fashion as it created a svc for the current master.

Just feels like a little bit of potentially wasted resources if they just sit and wait for master to go down if we can use them for reading.

**Reply by chakaz** [reply]
My recommendation would be to use the master for both reads and writes as much as possible. Dragonfly can do millions of QPS. Consider using the replica for reads only if the master truly and verifiably is overloaded.

**Comment by alexted** [comment]
Hello everyone! Tell me pls, has the situation changed in any way since the last post here?
At the moment, what capabilities does Dragonfly have for ensuring high availability and fault tolerance?

**Reply by romange** [reply]
Same like last year. It supports H/A via master/slave replication.

**Reply by alexted** [reply]
1. Does it allow for the creation of a full-fledged cluster?
2. What about multimaster? Do you plan to add support for this configuration?

**Reply by chakaz** [reply]
1. We have a helper script (see `cluster_mgr.py`) to do most basic setup, but we do not provide control and management facilities (although we do offer a managed service!)
2. We currently do not support, nor plan to support multi master in the near future
